src/framerate.rs

```diff
@@ -0,0 +1,265 @@
+    }
+
+    /// from_str is the private method we are going to use for parsing from both timebases and
+    /// framerates from strings.
+    fn from_str(value: &str, ntsc: Ntsc, is_base: bool) -> Result<Self, String> {
+        let mut result: Option<Result<Self, String>> = None;
+
+        if let Ok(parsed) = num::Rational64::from_str(value) {
+            return Self::from_rational(parsed, ntsc, is_base);
+        }
+
+        if let Ok(parsed) = i64::from_str(value) {
+            return Self::from_i64(parsed, ntsc, is_base);
+        }
+
+        if let Ok(parsed) = f64::from_str(value) {
+            result = Some(Self::from_f64(parsed, ntsc, is_base))
+        }
+
+        match result {
+            None => Err(format!(
+                "could not parse '{0}' as rational, int, or float for framerate.",
+                value
+            )),
+            Some(result) => result,
+        }
+    }
+
+    /// from_str_timebase creates a new framerate from a &str representing the timecode timebase
+    /// speed in frames-per-second.
+    ///
+    /// strings may be formatted as:
+    ///     - a fraction '24/1'
+    ///     - a float    '24.0'
+    ///     - an integer '24'
+    pub fn from_str_timebase(base: &str, ntsc: Ntsc) -> Result<Self, String> {
+        Self::from_str(base, ntsc, true)
+    }
+
+    /// from_str_timebase creates a new framerate from a &str representing the timecode timebase
+    /// speed in frames-per-second.
+    ///
+    /// strings may be formatted as:
+    ///     - a fraction '24000/1001'
+    ///     - a float    '23.98'
+    ///     - an integer '24'
+    pub fn from_str_playback(rate: &str, ntsc: Ntsc) -> Result<Self, String> {
+        Self::from_str(rate, ntsc, false)
+    }
+}
+
+impl fmt::Display for Framerate {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        let value_str = format!("{:.2}", self.value.to_f64().unwrap());
+        let mut value_str = value_str.trim_end_matches('0');
+        value_str = value_str.trim_end_matches('.');
+
+        let ntsc_str = match self.ntsc {
+            Ntsc::False() => "",
+            Ntsc::NonDropFrame() => " NTSC",
+            Ntsc::DropFrame() => " NTSC DF",
+        };
+        write!(f, "[{}{}]", value_str, ntsc_str)
+    }
+}
+
+const DROP_DIVISOR_PLAYBACK: num::Rational64 = num::Rational64::new_raw(30000, 1001);
+const DROP_DIVISOR_TIMEBASE: num::Rational64 = num::Rational64::new_raw(30, 1);
+
+/// validate_dropframe_ntsc validates that our ntsc and drop-frame settings are correct.
+fn validate_ntsc_value(value: num::Rational64, ntsc: Ntsc, is_base: bool) -> Result<(), String> {
+    // If this is not an NTSC value, return immediately.
+    if !ntsc.check() {
+        return Ok(());
+    }
+
+    if is_base {
+        // If this is a timebase, it needs to be a whole-number, or it is not NTSC compliant.
+        if !value.is_integer() {
+            return Err("ntsc timebases must be whole numbers".to_string());
```

Comment:
It may be nice to have error values that are not just strings, this can allow the user to check in code why something failed, but this is probably not a huge concern as most users will probably just print the error anyway.

Reply:
I'm still trying to figure out error handling in rust. Whst kind of type would you use for sentinel errors generally? A struct?

Reply:
I would do something like this:

```
pub enum ParsingError {
    NotWholeNumber,
    ...
}
```

Then make the error type of the result that enum.